autoload/vimlparser.vim

```diff
@@ -1865,6 +1905,22 @@ function! s:VimLParser.parse_lvalue()
   throw s:Err('Invalid Expression', node.pos)
 endfunction
 
+" TODO: merge with s:VimLParser.parse_lvalue()
+function! s:VimLParser.parse_constlvalue()
+  let p = s:LvalueParser.new(self.reader)
+  let node = p.parse()
+  if node.type == s:NODE_IDENTIFIER
+    if !s:isvarname(node.value)
+      throw s:Err(printf('E461: Illegal variable name: %s', node.value), node.pos)
+    endif
+  endif
+  if node.type == s:NODE_IDENTIFIER || node.type == s:NODE_CURLYNAME
```

Comment:
In contrast to `s:VimLParser.parse_lvalue()`, this does not handle

* `s:NODE_SUBSCRIPT` (`dict["key"]`)
* `s:NODE_SLICE` (`list[a:b]`)
* `s:NODE_DOT` (`dict.key`)
* `s:NODE_OPTION` (`&expandtab`)
* `s:NODE_ENV` (`$VAR`)
* `s:NODE_REG` (`@a`)

```
  if node.type == s:NODE_IDENTIFIER || node.type == s:NODE_CURLYNAME || node.type == s:NODE_SUBSCRIPT || node.type == s:NODE_SLICE || node.type == s:NODE_DOT || node.type == s:NODE_OPTION || node.type == s:NODE_ENV || node.type == s:NODE_REG
```

so `throw s:Err('Invalid Expression', node.pos)` will be executed.

Reply:
Do you have plan to sparate error message for E996?

```
let a = [0]
const a[0] = 1 " E996: Cannot lock a list or dict
let a = {}
const a["b"] = 1 " E996: Cannot lock a list or dict
const a.b = 1 " E996: Cannot lock a list or dict
const $a = 1 " E996: Cannot lock an environment variable
const @a = 1 " E996: Cannot lock a register
const &encoding = "foo" " E996: Cannot lock an option
```

Reply:
Sure. certainly it is better to throw E996 error here (from vimlparser).
I'll fix it.